### PR TITLE
Bug 993899 - To support Emoji in keyboard

### PR DESCRIPTION
**Pull request by jedireza**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=49b360a0f868e1eb9d0b43d89f8b710749de7542

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d4812c3ac0791fa83f432b05b7f4f9eb9ee37fe4

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=84debcd29a9ae1fa19d9451b0617bb6aaa23765c

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f3898f7b8057604daf20261bf5d5eaa5b6cca309

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1f83572e45a76a50520988ecf7fdfb67d9bbcd00

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e5db61eb226526ebbc6e4d15148ee9a9658db0b4

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c06287d21eed6e231b4a37229a9140454867a335

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0af15dc46cd4f0476f8a0878541ec107875682d5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5020dafb36ccaedb564d90b3fffcbeab954164d9

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=72696c3929ca20f9928927b43ba81ab3e943956f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=eace3d58bc146b348599496147a5519322756527

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=20a0033def55572172042593c248ab8a66fcb11b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bf0fd2fdc13ac0b6ffce709fb3636b6ee2d86e0f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5f63ebbd9449179f8abe6e2a419996637814d173

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e30bdac8dfcac8e37f67befc0a5e217cd6e37280

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=84d53b9cee7c79f6aeb75b984fbc8a4da7584092

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=72f388fab96129daa2231402d78197e2157d0702

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d6d85e267bc5266a4ee17180d6c066d68a92e8e5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2cf5d6182855d7da7226332265956cf3ed2a5172

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=267dc7cbb52c573ec858277507bc8698e3dd3266

File: apps/keyboard/style/keyboard.css [hunk 1]
```diff
@@ -827,7 +827,7 @@ bubble above the key when you tap and hold. */
 }
 
 .emoji {
-  font-family: sans-serif;
```
Comment (timdream) [inline apps/keyboard/style/keyboard.css]: Please remove this.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ce6c31780c5cc7c8d8979e11e69fef3423d02e9c

File: apps/keyboard/style/keyboard.css (hunk 1 above)
Comment (jedireza) [inline apps/keyboard/style/keyboard.css]: Fixed

File: apps/keyboard/js/layouts/emoji.js [hunk 2]
```diff
@@ -313,52 +387,56 @@ Keyboards.emoji = {
...
-                 { compositeKey: '💈', type: 'emoji' }, { compositeKey: '🚏', type: 'emoji' },
-                 { compositeKey: '🎫', type: 'emoji' }, { compositeKey: '🚦', type: 'emoji' },
-                 { compositeKey: '🚥', type: 'emoji' }, { compositeKey: '⚠', type: 'emoji' },
-                 { compositeKey: '🚧', type: 'emoji' }, { compositeKey: '🔰', type: 'emoji' },
-                 { compositeKey: '⛽', type: 'emoji' }, { compositeKey: '🏮', type: 'emoji' },
-                 { compositeKey: '🎰', type: 'emoji' }, { compositeKey: '♨', type: 'emoji' },
-                 { compositeKey: '🗿', type: 'emoji' }, { compositeKey: '🎪', type: 'emoji' },
-                 { compositeKey: '🎭', type: 'emoji' }, { compositeKey: '📍', type: 'emoji' },
-                 { compositeKey: '🚩', type: 'emoji' }
+    panelKeys: [{ compositeKey: '🇨', type: 'emoji' }, { compositeKey: '🇪', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: The flags are not encoded on it's right place. The U+1F1E8 along does not represent the flag of China -- the sequence of U+1F1E8 and U+1F1F3 would as shown below:

U+1F1E8 U+1F1F3
> 🇨🇳

U+1F1E8
> 🇨

U+1F1F3
> 🇳

Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: See also: https://en.wikipedia.org/wiki/Regional_Indicator_Symbol#Unicode_Block

Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Spain = ES

Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Or, for simple copy and paste: 🇪 🇸

File: apps/keyboard/js/layouts/emoji.js [hunk 3]
```diff
@@ -313,52 +387,56 @@ Keyboards.emoji = {
...
+                { compositeKey: '🇫', type: 'emoji' }, { compositeKey: '🇬', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: France: 🇫 🇷
UK: 🇬 🇧

File: apps/keyboard/js/layouts/emoji.js [hunk 4]
```diff
@@ -313,52 +387,56 @@ Keyboards.emoji = {
...
+                { compositeKey: '🇮', type: 'emoji' }, { compositeKey: '🇯', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Italy: 🇮 🇹
Japan: 🇯 🇵

File: apps/keyboard/js/layouts/emoji.js [hunk 5]
```diff
@@ -313,52 +387,56 @@ Keyboards.emoji = {
...
+                { compositeKey: '🇰', type: 'emoji' }, { compositeKey: '🇷', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Korea: 🇰 🇷
Russia: 🇷 🇺

File: apps/keyboard/js/layouts/emoji.js [hunk 6]
```diff
@@ -313,52 +387,56 @@ Keyboards.emoji = {
...
+                { compositeKey: '🇺', type: 'emoji' }, { compositeKey: '🌃', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: US: 🇺 🇸

File: apps/keyboard/js/layouts/emoji.js [hunk 7]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
       ]
     ]
     }, {     // Page 4 - Symbols
-    panelKeys: [ { compositeKey: '🆗', type: 'emoji' }, { compositeKey: '🔀', type: 'emoji' },
-                 { compositeKey: '🔁', type: 'emoji' }, { compositeKey: '🔂', type: 'emoji' },
-                 { compositeKey: '🆕', type: 'emoji' }, { compositeKey: '🆙', type: 'emoji' },
-                 { compositeKey: '🆒', type: 'emoji' }, { compositeKey: '🆓', type: 'emoji' },
-                 { compositeKey: '🆖', type: 'emoji' }, { compositeKey: '📶', type: 'emoji' },
-                 { compositeKey: '🎦', type: 'emoji' }, { compositeKey: '🈁', type: 'emoji' }
+    panelKeys: [{ compositeKey: '3', type: 'emoji' }, { compositeKey: '0', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Extra `3` key.

Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: I fail to understand why we need these plain numbers. Even if I try to force these code points shown in Emoji (by append a VS16), I still couldn't see an colored number.

Should we just remove these?

File: apps/keyboard/js/layouts/emoji.js [hunk 8]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
...
+                { compositeKey: '1', type: 'emoji' }, { compositeKey: '2', type: 'emoji' },
+                { compositeKey: '3', type: 'emoji' }, { compositeKey: '4', type: 'emoji' },
+                { compositeKey: '5', type: 'emoji' }, { compositeKey: '6', type: 'emoji' },
+                { compositeKey: '7', type: 'emoji' }, { compositeKey: '8', type: 'emoji' },
+                { compositeKey: '9', type: 'emoji' }, { compositeKey: '©', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: I don't see `©` in Firefox Emoji.

File: apps/keyboard/js/layouts/emoji.js [hunk 9]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
...
+                { compositeKey: '🚰', type: 'emoji' }, { compositeKey: '🚱', type: 'emoji' },
+                { compositeKey: '🚳', type: 'emoji' }, { compositeKey: '🚷', type: 'emoji' },
+                { compositeKey: '🚸', type: 'emoji' }, { compositeKey: '🚹', type: 'emoji' },
+                { compositeKey: '🚺', type: 'emoji' }, { compositeKey: '🚻', type: 'emoji' },
+                { compositeKey: '🚼', type: 'emoji' }, { compositeKey: '🚾', type: 'emoji' },
+                { compositeKey: '🛂', type: 'emoji' }, { compositeKey: '🛃', type: 'emoji' },
+                { compositeKey: '🛄', type: 'emoji' }, { compositeKey: '🛅', type: 'emoji' },
+                { compositeKey: '\u203c\ufe0f', type: 'emoji' }, { compositeKey: '\u2049\ufe0f', type: 'emoji' },
+                { compositeKey: '\u2122\ufe0f', type: 'emoji' }, { compositeKey: '\u2139\ufe0f', type: 'emoji' },
+                { compositeKey: '\u2194\ufe0f', type: 'emoji' }, { compositeKey: '\u2195\ufe0f', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: I don't see `™` in Firefox Emoji

File: apps/keyboard/js/layouts/emoji.js [hunk 10]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
...
+                { compositeKey: '🚮', type: 'emoji' }, { compositeKey: '🚯', type: 'emoji' },
+                { compositeKey: '🚰', type: 'emoji' }, { compositeKey: '🚱', type: 'emoji' },
+                { compositeKey: '🚳', type: 'emoji' }, { compositeKey: '🚷', type: 'emoji' },
+                { compositeKey: '🚸', type: 'emoji' }, { compositeKey: '🚹', type: 'emoji' },
+                { compositeKey: '🚺', type: 'emoji' }, { compositeKey: '🚻', type: 'emoji' },
+                { compositeKey: '🚼', type: 'emoji' }, { compositeKey: '🚾', type: 'emoji' },
+                { compositeKey: '🛂', type: 'emoji' }, { compositeKey: '🛃', type: 'emoji' },
+                { compositeKey: '🛄', type: 'emoji' }, { compositeKey: '🛅', type: 'emoji' },
+                { compositeKey: '\u203c\ufe0f', type: 'emoji' }, { compositeKey: '\u2049\ufe0f', type: 'emoji' },
+                { compositeKey: '\u2122\ufe0f', type: 'emoji' }, { compositeKey: '\u2139\ufe0f', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: I don't see `™` in Firefox Emoji

File: apps/keyboard/js/layouts/emoji.js [hunk 11]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
...
+                { compositeKey: '\u2194\ufe0f', type: 'emoji' }, { compositeKey: '\u2195\ufe0f', type: 'emoji' },
+                { compositeKey: '\u2196\ufe0f', type: 'emoji' }, { compositeKey: '\u2197\ufe0f', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: And these.

File: apps/keyboard/js/layouts/emoji.js [hunk 12]
```diff
@@ -403,12 +475,104 @@ Keyboards.emoji = {
...
+                { compositeKey: '\u2198\ufe0f', type: 'emoji' }, { compositeKey: '\u21a9\ufe0f', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: and U+2198

File: apps/keyboard/js/layouts/emoji.js (hunk 7 above)
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: Are you trying to target 1⃣, which is U+0031 U+20E3 ? If so these keys should be defined as such and the font should encode the glyph as such.

Neither is correctly done currently.

Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: I can also see a Emoji glyph a number character follow by a VS16 on Apple Color Emoji, if you want to support it you could do that too (but VS16 has to included in the key definition, like other keys).

**Comment by jedireza** [comment]
@pivanov are you able to make these changes?

**Comment by pivanov** [comment]
@jedireza @timdream I asked Jonathan how we can use VS16 in Emoji Font because now we use something like:
code point u2198 use these glyphs
and also I'm not sure how to use two code points like in country flags ... waiting Jonathan's replay

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2533eb755e2aa836f31e122d07ada2887061b12f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fa9906ddb9c859c8f2f63ccaa54a3e61c9b41b56

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=87a853ea00095b9cd499058f1a57ad2039819c42

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8a0bc740e643cc8d22382d8d7359bed98e1ff397

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=017fb4c51f78a01dc8192931a025a1b73ea04c47

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0ad16fc3da360aa8dff2348c198ba7a4fdbe660f

File: apps/keyboard/js/layouts/emoji.js [hunk 13]
```diff
@@ -196,79 +232,121 @@ Keyboards.emoji = {
...
-                { compositeKey: '🎅', type: 'emoji' }, { compositeKey: '🎄', type: 'emoji' },
-                { compositeKey: '🎁', type: 'emoji' }, { compositeKey: '🎋', type: 'emoji' },
+    panelKeys: [{ compositeKey: '🀄', type: 'emoji' }, { compositeKey: '🃏', type: 'emoji' },
+                { compositeKey: '🌽', type: 'emoji' }, { compositeKey: '🍅', type: 'emoji' },
+                { compositeKey: '🍆', type: 'emoji' }, { compositeKey: '🍇', type: 'emoji' },
+                { compositeKey: '🍈', type: 'emoji' }, { compositeKey: '🍉', type: 'emoji' },
+                { compositeKey: '🍊', type: 'emoji' }, { compositeKey: '🍋', type: 'emoji' },
+                { compositeKey: '🍌', type: 'emoji' }, { compositeKey: '🍍', type: 'emoji' },
+                { compositeKey: '🍎', type: 'emoji' }, { compositeKey: '🍏', type: 'emoji' },
+                { compositeKey: 'ἵ', type: 'emoji' }, { compositeKey: '🍐', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: What is `ἵ`?

File: apps/keyboard/js/layouts/emoji.js [hunk 14]
```diff
@@ -403,12 +473,98 @@ Keyboards.emoji = {
       ]
     ]
     }, {     // Page 4 - Symbols
-    panelKeys: [ { compositeKey: '🆗', type: 'emoji' }, { compositeKey: '🔀', type: 'emoji' },
-                 { compositeKey: '🔁', type: 'emoji' }, { compositeKey: '🔂', type: 'emoji' },
-                 { compositeKey: '🆕', type: 'emoji' }, { compositeKey: '🆙', type: 'emoji' },
-                 { compositeKey: '🆒', type: 'emoji' }, { compositeKey: '🆓', type: 'emoji' },
-                 { compositeKey: '🆖', type: 'emoji' }, { compositeKey: '📶', type: 'emoji' },
-                 { compositeKey: '🎦', type: 'emoji' }, { compositeKey: '🈁', type: 'emoji' }
+    panelKeys: [{ compositeKey: '🅰', type: 'emoji' }, { compositeKey: '🅱', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `{ compositeKey: '🅰\ufe0f', type: 'emoji' }, { compositeKey: '🅱\ufe0f', type: 'emoji' },`

File: apps/keyboard/js/layouts/emoji.js [hunk 15]
```diff
@@ -403,12 +473,98 @@ Keyboards.emoji = {
...
+                { compositeKey: '🅾', type: 'emoji' }, { compositeKey: '🅿', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `{ compositeKey: '🅾\ufe0f', type: 'emoji' }, { compositeKey: '🅿\ufe0f', type: 'emoji' },`

File: apps/keyboard/js/layouts/emoji.js [hunk 16]
```diff
@@ -403,12 +473,98 @@ Keyboards.emoji = {
...
+                { compositeKey: '🆎', type: 'emoji' }, { compositeKey: '🆑', type: 'emoji' },
+                { compositeKey: '🆒', type: 'emoji' }, { compositeKey: '🆓', type: 'emoji' },
+                { compositeKey: '🆔', type: 'emoji' }, { compositeKey: '🆕', type: 'emoji' },
+                { compositeKey: '🆖', type: 'emoji' }, { compositeKey: '🆗', type: 'emoji' },
+                { compositeKey: '🆘', type: 'emoji' }, { compositeKey: '🆙', type: 'emoji' },
+                { compositeKey: '🆚', type: 'emoji' }, { compositeKey: '🈁', type: 'emoji' },
+                { compositeKey: '🈂', type: 'emoji' }, { compositeKey: '🈚', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `{ compositeKey: '🈂\ufe0f', type: 'emoji' }, { compositeKey: '🈚\ufe0f', type: 'emoji' },`

File: apps/keyboard/js/layouts/emoji.js [hunk 17]
```diff
@@ -403,12 +473,98 @@ Keyboards.emoji = {
...
+                { compositeKey: '🈯', type: 'emoji' }, { compositeKey: '🈲', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `{ compositeKey: '🈯\ufe0f', type: 'emoji' }, { compositeKey: '🈲', type: 'emoji' },`

File: apps/keyboard/js/layouts/emoji.js [hunk 18]
```diff
@@ -403,12 +473,98 @@ Keyboards.emoji = {
...
+                { compositeKey: '🈳', type: 'emoji' }, { compositeKey: '🈴', type: 'emoji' },
+                { compositeKey: '🈵', type: 'emoji' }, { compositeKey: '🈶', type: 'emoji' },
+                { compositeKey: '🈷', type: 'emoji' }, { compositeKey: '🈸', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `{ compositeKey: '🈷\ufe0f', type: 'emoji' }, { compositeKey: '🈸', type: 'emoji' },`

File: apps/keyboard/js/layouts/emoji.js [hunk 19]
```diff
@@ -196,79 +232,121 @@ Keyboards.emoji = {
...
     ]}, {    // Page 2 - Object
-    panelKeys: [{ compositeKey: '🎍', type: 'emoji' }, { compositeKey: '💝', type: 'emoji' },
-                { compositeKey: '🎎', type: 'emoji' }, { compositeKey: '🎒', type: 'emoji' },
-                { compositeKey: '🎓', type: 'emoji' }, { compositeKey: '🎏', type: 'emoji' },
-                { compositeKey: '🎆', type: 'emoji' }, { compositeKey: '🎇', type: 'emoji' },
-                { compositeKey: '🎐', type: 'emoji' }, { compositeKey: '🎑', type: 'emoji' },
-                { compositeKey: '🎃', type: 'emoji' }, { compositeKey: '👻', type: 'emoji' },
-                { compositeKey: '🎅', type: 'emoji' }, { compositeKey: '🎄', type: 'emoji' },
-                { compositeKey: '🎁', type: 'emoji' }, { compositeKey: '🎋', type: 'emoji' },
+    panelKeys: [{ compositeKey: '🀄', type: 'emoji' }, { compositeKey: '🃏', type: 'emoji' },
```
Comment (timdream) [inline apps/keyboard/js/layouts/emoji.js]: `panelKeys: [{ compositeKey: '🀄\ufe0f', type: 'emoji' }, { compositeKey: '🃏', type: 'emoji' },`

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ac35e28fad5cd1dceff4b219dc6f61447863cef3

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5f6b7326f27538623b28474f8ffbc10ae4da397c

**Pull request merged**